File: test/python/test_sub.py
#!/usr/bin/env python

import sys
import json
import numpy as np
import time

sys.path.append('../../build')
sys.path.append('../../build/swig')
import quickmsg_py

class SubscriberImpl(quickmsg_py.AsyncSubscriber):
    def __init__(self, *args, **kwargs):
        super(SubscriberImpl, self).__init__(*args, **kwargs)

    def subscriber_impl(self, msg):
        print 'Python inherited subscriber callback'
        print 'got message', msg.msg

if __name__=='__main__':
    s = SubscriberImpl('chatter')
    s.spin()
    

File: test/cpp/test_sub.cpp
#include <quickmsg/quickmsg.hpp>
#include <sstream>
#include <iostream>
#include <czmq.h>
#include <zyre.h>

namespace qm = quickmsg;

class SubImpl : public qm::AsyncSubscriber
{
public:

  SubImpl(const std::string& topic) : qm::AsyncSubscriber(topic) {}
  virtual ~SubImpl() {}
  virtual void handle_message(const qm::Message* msg)
  {
    std::cout<<"inherited impl"<<std::endl;
    std::cout << "msg: " << msg->msg << std::endl;
  }
};

void
handler(const qm::Message* msg, void* args)
{
  std::cout << "Message received: " << msg->msg << std::endl;
}

// void qmshutdown()
// {
//   qm::shutdown(": done");
// }

int
main(int argc, char** argv)
{
  //std::atexit(qmshutdown);
  qm::init("test_cpp_sub");
  qm::AsyncSubscriber* async_sub = new qm::AsyncSubscriber("chatter", handler, NULL);
  async_sub->spin();
  delete async_sub;
  qm::shutdown(": done");
  //  qm::Subscriber sub("chatter", 20);
  // SubImpl sub("chatter");
  // while (qm::ok())
  // {
  //   sleep(1);
  // }
  return 0;
}
